teps/0041-tekton-component-versioning.md

```diff
@@ -0,0 +1,346 @@
+install/upgrade/downgrade/reinstall a resource from catalog only if the
+newer resource version is compatible with Pipelines.
+
+## Motivation
+
+Finding the version of tekton components such as Tekton Pipelines etc which
+are currently installed on the cluster is not an easy thing specially when we
+may not have specific permissions to read the deployment and fetch the
+version. In the current scenario we can fetch version from the controller
+deployment and for that we need:-
+
+1. Namespace in which tekton component is installed
+2. Permissions to view deployment/pod in installed namespace
+
+The above permission may not be granted to all the users and finding the version
+becomes difficult in that case. To make `Tekton Hub CLI` subcommands such as
+install, upgrade etc work efficiently it needs to know the installed Tekton
+Component version.
+
+### Goals
+
+1. All the users having access to the cluster should be able to view the version
+   of Tekton component installed using Tekton CLI.
+2. Tekton Hub CLI should be able to install/upgrade resources from catalog on the
+   basis of Tekton component installed.
+
+### Non-Goals
+
+1. How to handle versioning of components which were released prior to this
+   TEP will not be covered.
+2. Maintaining the agreed upon solution in the respective components will not
+   be covered.
+
+### Use Cases (optional)
+
+#### Tekton Hub CLI
+
+1. Tekton Hub CLI will provide an install command like `tkn hub install task buildah`.
+2. Tekton Hub CLI will provide an update command like `tkn hub upgrade task buildah`.
+3. Tekton Hub CLI will provide an check-updates command to list all tasks whose newer
+   version is available.
+
+We want to perform above operations using `tkn hub` CLI based on the version
+of Tekton Pipelines installed on the cluster and `pipelines.minVersion`
+annotation available in catalog like [this](https://github.com/tektoncd/catalog/blob/master/task/buildah/0.2/buildah.yaml#L9).
+This feature can later be extended for the other Tekton components such
+as Triggers, Dashboard etc.
+
+#### New Tekton User
+
+New Tekton users can get to know the version of Tekton components
+installed so that they can create their pipelines based on the features
+available in that version.
+
+## Requirements
+
+None
+
+## Proposal
+
+This proposal is to allow the system authenticated users of the cluster
+to access the version of the Tekton component installed even if they don't
+have proper rights to read the version from the controller deployments.
+
+The proposed solution is to create a `ConfigMap` and to make that readable
+to any authenticated user through a dedicated `Role` and `RoleBinding`.
+Design details are present in [Design Details](#design-details).
+
+### Risks and Mitigations
+
+1. Tekton CLI will not be able to access the version information from previously
+   released tekton components without this functionality.
+
+2. Tekton CLI needs to handle backward compatibility,
```

Comment:
does backward compatibility here mean the CLI needs to be able to read the versions of previously released components? Or does this mean it needs to gracefully error out when trying to read the version of previously released components? either way i think this point would be clearer with a bit more detail